csrc/kernel_cache.cpp

```diff
@@ -376,9 +376,8 @@ void prepareRuntimeOrder(
     available_input.insert(input_val);
 
     if (auto input_tv = dynamic_cast<TensorView*>(input_val)) {
-      auto root_dom = TensorDomain::noReductions(input_tv->getRootDomain());
-      for (const size_t dim : c10::irange(root_dom.size())) {
-        const auto extent = root_dom[dim]->getMaybeExpandedExtent();
+      for (auto id : TensorDomain::noReductions(input_tv->getLeafDomain())) {
```

Comment:
I'm hitting an assert without this change, mainly because the NamedScalar "gridDim.x" is not contained in the root domain but is only in the leaf.
I suspect we never encountered this issue before because the segmenter was always used before the scheduler, but in `SchedulingMode::ReductionScheduler` it is used after.
Also, in theory, I think that leaf domain makes more sense here because this is indeed the domains tha the kernel sees to creates the loop, but I might be wrong. Let me know your thoughts

Reply:
but it might that in theory we need both leaf and root domain, but at least we don't encounter such a case

Reply:
Just a clarification so I can understand better. There are two pathways
(1) Automatic scheduling using scheduleReduction and FusionExecutor: this required careful modification of the reduction scheduler.
(2) Fully automated segmentation+scheduling using FusionExecutorCache: which is our ultimate target.

Is this change needed for both or just one of the paths?

Reply:
This change is needed only for path (1). I would say this is part of additional "book keeping" we need to handle -- in this case it is handling the named scalars `gridDim` and `blockDim`

Reply:
> I suspect we never encountered this issue before because the segmenter was always used before the scheduler, but in SchedulingMode::ReductionScheduler it is used after.

Yes, very likely. 

> but it might that in theory we need both leaf and root domain, but at least we don't encounter such a case

@naoyam , wdyt? Does it make more sense to add the rfactor domain instead? AFAIU, this code is trying to add input dependencies for consuming ops. Those use rfactor (not leaf, unless it happens to be the same) extents of their producers.

Reply:
What assert is this?

> I'm hitting an assert without this change

Reply:
Could you please give more details why that assertion fails too?

Reply:
> What assert is this?
> 
> > I'm hitting an assert without this change

It's the assert at the end of that function body, `csrc/kernel_cache.cpp:420`

This assert makes sense, since `gridDim` is an input of the group, that dependency needs to be satisfied. That's why we need to add this extent in `available_input`, and therefore iterate on the leaf domain.

However, let me point out that if we decide to remove `gridDim` from the segmented group's input (as suggested in https://github.com/NVIDIA/Fuser/pull/1853#discussion_r1537941767), this modification will probably not be needed anymore.